internal/pkg/cosign/sign.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cosign
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"encoding/base64"
+	"io"
+
+	"github.com/sigstore/cosign/pkg/oci"
+	"github.com/sigstore/cosign/pkg/oci/static"
+	"github.com/sigstore/sigstore/pkg/signature"
+	signatureoptions "github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+type Signer interface {
+	Sign(ctx context.Context, payload io.Reader) (oci.Signature, crypto.PublicKey, error)
+}
+
+// PayloadSigner implements `Signer`
+type PayloadSigner struct {
+	PayloadSigner         signature.Signer
+	PayloadSignerOpts     []signature.SignOption
+	PublicKeyProviderOpts []signature.PublicKeyOption
+}
+
+// Sign uses the PayloadSigner to sign the requested payload, then returns the signature, the public key associated with it, the signed payload
```

Comment:
I'd move this onto `PayloadSigner` since most folks will see the interface comment via things like IDEs when looking at callsites to this method.

For interface impls, I always use: `// {function} implements {interface}`